### PR TITLE
Adding a few useful helper functions

### PR DESCRIPTION
**Pull request by benjaminsavage**
We will definitely need these two helper functions at the next stage. I've added a few tests for them that illustrate how they will be used and ensure the expected behavior.

### CODE REVIEW

File: ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs [hunk 1]
```diff
@@ -92,18 +95,52 @@ impl<F: PrimeField, const λ: usize, const P: usize, const M: usize> ProofGenera
 
         output
     }
+
+    fn gen_proof_shares_from_prss<C>(ctx: &C, record_counter: &mut RecordId) -> ([F; P], [F; P])
```
Comment (cberkhoff) [inline ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs:99]: I think it would be good to add documentation to ProofGenerator.

**Review by danielmasny — APPROVED** [review]
That works for me

**Comment by andyleiserson** [comment]
Is there a way to use `left` and `right` instead of `.0` and `.1`? I understand not wanting to use `Replicated` / `AdditiveShare` since that's not really what these are. Maybe return a simple helper type something like `struct LeftRight<T> { left: T, right: T }` instead of a tuple? Or just deconstruct the return value as `(foo_left, foo_right)` immediately.

It might be possible to use the `StdArray` wrapper type for the proofs (instead of `[F; P]`), which would provide pre-built support for generation with PRSS, subtraction for `gen_other_proof_share`, and sending over channels. Maybe `ShareArray` is the better name for `StdArray` that I wish I had thought of earlier.

**Comment by benjaminsavage** [comment]
These are all great suggestions. @danielmasny and I can follow up with them.

**Comment by benjaminsavage** [comment]
@andyleiserson - I started trying to use StdArray, but I got stuck along the way. It immediately started spreading and before long the whole file needed StdArray throughout, not just in this one function.

Changing `gen_other_proof_share` was easy, but then I needed to change `gen_artefacts_from_recursive_step` which meant changing `gen_challenge_and_recurse` to accept StdArray, and I couldn't make it work there because `compute_hash` wants something it can iterate through and serialize, and I got stuck there.

In general though, are we sure there's a big benefit to this? Vanilla arrays are nice a readable, and the only benefit we seem to get from StdArray is that you *can't* access the elements.

**Comment by andyleiserson** [comment]
I think it would be okay to add some things like `From`, `Into`, and `Deref` to `StdArray`. It wasn't needed for vectorization, but if we're using it for other things, as you found, it's hard to do much with it without some conversions.

Adding them to StdArray doesn't directly make them available to vectorized code in most cases, because the vectorized code is using the `SharedValueArray` trait rather than `StdArray` directly.

**Comment by andyleiserson** [comment]
This is what I had in mind: https://github.com/private-attribution/ipa/commit/a9587f827189ed4aaba1afdb4f47da563230042e

**Pull request merged**